File: tools/gnark/internal/abi/note_reshape_witness_binary.go
package abi

import (
	"bytes"
	"fmt"

	"github.com/mizufinance/shieldd/tools/gnark/internal/generated"
)

const (
	noteReshapeWitnessMagic   = "PNWG"
	noteReshapeWitnessVersion = 6
	maxNoteReshapeItems       = 8
)

type NoteReshapeSpendWitnessV6Binary struct {
	IsDummy                   bool
	Nullifier                 [32]byte
	DummyNullifierSeed        [32]byte
	SpentNoteBlinding         [32]byte
	SpentNoteAmount           [32]byte
	StateCommitmentCommitment [32]byte
	StateCommitmentPosition   uint64
	StateCommitmentAuthPath   [][3][32]byte
	SpendAuthRandomizer       [32]byte
	RKAffine                  PointAffineBinary
	HistoryRequired           bool
}

type NoteReshapeOutputWitnessV6Binary struct {
	NoteCommitment      [32]byte
	CreatedNoteBlinding [32]byte
	CreatedNoteAmount   [32]byte
}

type NoteReshapeSharedNoteContextWitnessV6Binary struct {
	AssetID [32]byte
	DivGen  PointAffineBinary
}

type NoteReshapeWitnessV6Binary struct {
	TotalLength              uint32
	FamilyID                 uint32
	NIn                      uint32
	NOut                     uint32
	Anchor                   [32]byte
	ClaimedStatementHash     [32]byte
	AssetAnchor              [32]byte
	ComplianceAnchor         [32]byte
	RoutingTag               [32]byte
	RoutingParameterSetID    [32]byte
	RecentPositionFloor      [32]byte
	ActionBalanceBlinding    [32]byte
	NK                       [32]byte
	AssetPath                MerklePathBinary
	AssetPosition            uint64
	AssetIndexedLeaf         IndexedLeafBinary
	AssetIndexedLeafDKPub    PointAffineBinary
	AssetIndexedLeafRingPK   PointAffineBinary
	IsRegulated              bool
	RegulatedPrecision       uint8
	UnregulatedPrecision     uint8
	RoutingAsOfHeight        uint64
	RoutingNonce             [32]byte
	SenderCompliancePath     MerklePathBinary
	SenderCompliancePosition uint64
	SenderSlotID             [32]byte
	SenderSlotDerivation     [32]byte
	SenderD                  [32]byte
	SenderStatus             [32]byte
	Shared                   NoteReshapeSharedNoteContextWitnessV6Binary
	Spends                   []NoteReshapeSpendWitnessV6Binary
	Outputs                  []NoteReshapeOutputWitnessV6Binary
	BalanceCommitmentAffine  PointAffineBinary
	AKAffine                 PointAffineBinary
}

func DecodeNoteReshapeWitnessV6(payload []byte) (*NoteReshapeWitnessV6Binary, generated.NoteReshapeFamilySpec, error) {
	witness, err := decodeNoteReshapeWitnessV6(payload)
	if err != nil {
		return nil, generated.NoteReshapeFamilySpec{}, err
	}
	family, ok := generated.NoteReshapeFamilyByID(witness.FamilyID)
	if !ok {
		return nil, generated.NoteReshapeFamilySpec{}, fmt.Errorf("unknown note reshape family id %d", witness.FamilyID)
	}
	if int(witness.NIn) != family.NIn || int(witness.NOut) != family.NOut {
		return nil, generated.NoteReshapeFamilySpec{}, fmt.Errorf("note reshape witness shape mismatch: got %dx%d, expected %dx%d", witness.NIn, witness.NOut, family.NIn, family.NOut)
	}
	return witness, family, nil
}

func decodeNoteReshapeWitnessV6(payload []byte) (*NoteReshapeWitnessV6Binary, error) {
	reader := bytes.NewReader(payload)
	magic, err := readExact(reader, 4)
	if err != nil {
		return nil, err
	}
	if string(magic) != noteReshapeWitnessMagic {
		return nil, fmt.Errorf("invalid note reshape witness magic %q", string(magic))
	}
	version, err := readU32(reader)
	if err != nil {
		return nil, err
	}
	if version != noteReshapeWitnessVersion {
		return nil, fmt.Errorf("unsupported note reshape witness version %d", version)
	}
	totalLength, err := readU32(reader)
	if err != nil {
		return nil, err
	}
	if totalLength != uint32(len(payload)) {
		return nil, fmt.Errorf("payload length mismatch: header=%d actual=%d", totalLength, len(payload))
	}
	witness := &NoteReshapeWitnessV6Binary{TotalLength: totalLength}
	if witness.FamilyID, err = readU32(reader); err != nil {
		return nil, err
	}
	if witness.NIn, err = readU32(reader); err != nil {
		return nil, err
	}
	if witness.NOut, err = readU32(reader); err != nil {
		return nil, err
	}
	family, ok := generated.NoteReshapeFamilyByID(witness.FamilyID)
	if !ok {
		return nil, fmt.Errorf("unknown note reshape family id %d", witness.FamilyID)
	}
	if witness.Anchor, err = read32(reader); err != nil {
		return nil, err
	}
	if witness.ClaimedStatementHash, err = read32(reader); err != nil {
		return nil, err
	}
	if witness.AssetAnchor, err = read32(reader); err != nil {
		return nil, err
	}
	if witness.ComplianceAnchor, err = read32(reader); err != nil {
		return nil, err
	}
	if witness.RoutingTag, err = read32(reader); err != nil {
		return nil, err
	}
	if witness.RoutingParameterSetID, err = read32(reader); err != nil {
		return nil, err
	}
	if witness.RecentPositionFloor, err = read32(reader); err != nil {
		return nil, err
	}
	if witness.ActionBalanceBlinding, err = readFr32(reader); err != nil {
		return nil, err
	}
	if witness.NK, err = read32(reader); err != nil {
		return nil, err
	}
	if witness.AssetPath, err = readMerklePath(reader); err != nil {
		return nil, err
	}
	if witness.AssetPosition, err = readU64(reader); err != nil {
		return nil, err
	}
	if witness.AssetIndexedLeaf, err = readIndexedLeaf(reader); err != nil {
		return nil, err
	}
	if witness.AssetIndexedLeafDKPub, err = readPointAffine(reader); err != nil {
		return nil, err
	}
	if witness.AssetIndexedLeafRingPK, err = readPointAffine(reader); err != nil {
		return nil, err
	}
	if witness.IsRegulated, err = readBool(reader); err != nil {
		return nil, err
	}
	if witness.RegulatedPrecision, err = readU8(reader); err != nil {
		return nil, err
	}
	if witness.UnregulatedPrecision, err = readU8(reader); err != nil {
		return nil, err
	}
	if witness.RoutingAsOfHeight, err = readU64(reader); err != nil {
		return nil, err
	}
	if witness.RoutingNonce, err = read32(reader); err != nil {
		return nil, err
	}
	if witness.SenderCompliancePath, err = readMerklePath(reader); err != nil {
		return nil, err
	}
	if witness.SenderCompliancePosition, err = readU64(reader); err != nil {
		return nil, err
	}
	if witness.SenderSlotID, err = read32(reader); err != nil {
		return nil, err
	}
	if witness.SenderSlotDerivation, err = read32(reader); err != nil {
		return nil, err
	}
	if witness.SenderD, err = read32(reader); err != nil {
		return nil, err
	}
	if witness.SenderStatus, err = read32(reader); err != nil {
		return nil, err
	}
	if witness.Shared.AssetID, err = read32(reader); err != nil {
		return nil, err
	}
	if witness.Shared.DivGen, err = readPointAffine(reader); err != nil {
		return nil, err
	}
	if witness.NIn > maxNoteReshapeItems || witness.NOut > maxNoteReshapeItems {
		return nil, fmt.Errorf("note reshape witness exceeds maximum item count")
	}
	witness.Spends = make([]NoteReshapeSpendWitnessV6Binary, witness.NIn)
	for i := range witness.Spends {
		if witness.Spends[i], err = readNoteReshapeSpend(reader, family.InputPadding == generated.InputPaddingSyntheticPrivate); err != nil {
			return nil, err
		}
	}
	witness.Outputs = make([]NoteReshapeOutputWitnessV6Binary, witness.NOut)
	for i := range witness.Outputs {
		if witness.Outputs[i], err = readNoteReshapeOutput(reader); err != nil {
			return nil, err
		}
	}
	if witness.BalanceCommitmentAffine, err = readPointAffine(reader); err != nil {
		return nil, err
	}
	if witness.AKAffine, err = readPointAffine(reader); err != nil {
		return nil, err
	}
	if reader.Len() != 0 {
		return nil, fmt.Errorf("trailing bytes in note reshape witness: %d", reader.Len())
	}
	return witness, nil
}

func readNoteReshapeSpend(reader *bytes.Reader, syntheticPrivatePadding bool) (NoteReshapeSpendWitnessV6Binary, error) {
	var out NoteReshapeSpendWitnessV6Binary
	var err error
	if syntheticPrivatePadding {
		flag, readErr := readU32(reader)
		if readErr != nil {
			return out, readErr
		}
		if flag > 1 {
			return out, fmt.Errorf("invalid note reshape input dummy flag %d", flag)
		}
		out.IsDummy = flag == 1
		if out.DummyNullifierSeed, err = read32(reader); err != nil {
			return out, err
		}
	}
	if out.Nullifier, err = read32(reader); err != nil {
		return out, err
	}
	if out.SpentNoteBlinding, err = read32(reader); err != nil {
		return out, err
	}
	if out.SpentNoteAmount, err = read32(reader); err != nil {
		return out, err
	}
	if out.StateCommitmentCommitment, err = read32(reader); err != nil {
		return out, err
	}
	if out.StateCommitmentPosition, err = readU64(reader); err != nil {
		return out, err
	}
	if out.StateCommitmentAuthPath, err = readTriplePath(reader); err != nil {
		return out, err
	}
	if out.SpendAuthRandomizer, err = readFr32(reader); err != nil {
		return out, err
	}
	if out.RKAffine, err = readPointAffine(reader); err != nil {
		return out, err
	}
	if out.HistoryRequired, err = readBool(reader); err != nil {
		return out, err
	}
	return out, nil
}

func readNoteReshapeOutput(reader *bytes.Reader) (NoteReshapeOutputWitnessV6Binary, error) {
	var out NoteReshapeOutputWitnessV6Binary
	var err error
	if out.NoteCommitment, err = read32(reader); err != nil {
		return out, err
	}
	if out.CreatedNoteBlinding, err = read32(reader); err != nil {
		return out, err
	}
	if out.CreatedNoteAmount, err = read32(reader); err != nil {
		return out, err
	}
	return out, nil
}

File: crates/core/component/shielded-pool/src/gnark/note_reshape_witness_binary.rs
use anyhow::{bail, Context, Result};

use crate::gnark::{
    binary::{encode_triple_path_32, put_bytes, put_u32, put_u64, put_u8, BinaryCursor},
    note_reshape_witness::{
        NoteReshapeOutputWitnessV6, NoteReshapeSharedNoteContextWitnessV6,
        NoteReshapeSpendWitnessV6, NoteReshapeWitnessV6,
    },
    typed::{decode_indexed_leaf, encode_indexed_leaf, encode_merkle_path, encode_point_affine},
};

const NOTE_RESHAPE_WITNESS_MAGIC: &[u8; 4] = b"PNWG";
const NOTE_RESHAPE_WITNESS_VERSION: u32 = 8;

impl NoteReshapeWitnessV6 {
    pub fn encode(&self) -> Result<Vec<u8>> {
        let mut buf = Vec::new();
        put_bytes(&mut buf, NOTE_RESHAPE_WITNESS_MAGIC);
        put_u32(&mut buf, NOTE_RESHAPE_WITNESS_VERSION);
        put_u32(&mut buf, 0);
        put_u32(&mut buf, self.family_id.get());
        put_u32(&mut buf, self.n_in);
        put_u32(&mut buf, self.n_out);
        put_bytes(&mut buf, &self.anchor);
        put_bytes(&mut buf, &self.claimed_statement_hash);
        put_bytes(&mut buf, &self.asset_anchor);
        put_bytes(&mut buf, &self.compliance_anchor);
        put_bytes(&mut buf, &self.routing_tag);
        put_bytes(&mut buf, &self.routing_parameter_set_id);
        put_bytes(&mut buf, &self.recent_position_floor);
        put_bytes(&mut buf, &self.action_balance_blinding);
        put_bytes(&mut buf, &self.nk);
        encode_merkle_path(&mut buf, &self.asset_path)?;
        put_u64(&mut buf, self.asset_position);
        encode_indexed_leaf(&mut buf, &self.asset_indexed_leaf);
        encode_point_affine(&mut buf, &self.asset_indexed_leaf_dk_pub_affine);
        encode_point_affine(&mut buf, &self.asset_indexed_leaf_ring_pk_affine);
        put_u8(&mut buf, u8::from(self.is_regulated));
        put_u8(&mut buf, self.regulated_precision);
        put_u8(&mut buf, self.unregulated_precision);
        put_u64(&mut buf, self.routing_as_of_height);
        put_bytes(&mut buf, &self.routing_nonce);
        encode_merkle_path(&mut buf, &self.sender_compliance_path)?;
        put_u64(&mut buf, self.sender_compliance_position);
        put_bytes(&mut buf, &self.sender_d);
        put_bytes(&mut buf, &self.sender_status);
        put_bytes(&mut buf, &self.shared.asset_id);
        encode_point_affine(&mut buf, &self.shared.diversified_generator_affine);
        for spend in &self.spends {
            encode_spend(
                &mut buf,
                spend,
                self.family_id.spec().input_padding
                    == crate::note_reshape::InputPaddingPolicy::SyntheticPrivate,
            )?;
        }
        for output in &self.outputs {
            encode_output(&mut buf, output);
        }
        encode_point_affine(&mut buf, &self.balance_commitment_affine);
        encode_point_affine(&mut buf, &self.ak_affine);

        let total_len =
            u32::try_from(buf.len()).context("encoded note reshape witness exceeds u32")?;
        buf[8..12].copy_from_slice(&total_len.to_le_bytes());
        Ok(buf)
    }

    pub fn decode(bytes: &[u8]) -> Result<Self> {
        let mut cursor = BinaryCursor::new(bytes);
        if cursor.read_fixed::<4>()? != *NOTE_RESHAPE_WITNESS_MAGIC {
            bail!("invalid note reshape witness magic")
        }
        let version = cursor.read_u32()?;
        if version != NOTE_RESHAPE_WITNESS_VERSION {
            bail!("unsupported note reshape witness version {version}")
        }
        let total_length = cursor.read_u32()?;
        if total_length as usize != bytes.len() {
            bail!(
                "note reshape witness length mismatch: header={}, actual={}",
                total_length,
                bytes.len()
            )
        }
        let family_id = crate::NoteReshapeFamilyId::try_from(cursor.read_u32()?)?;
        let n_in = cursor.read_u32()?;
        let n_out = cursor.read_u32()?;
        let anchor = cursor.read_fixed::<32>()?;
        let claimed_statement_hash = cursor.read_fixed::<32>()?;
        let asset_anchor = cursor.read_fixed::<32>()?;
        let compliance_anchor = cursor.read_fixed::<32>()?;
        let routing_tag = cursor.read_fixed::<32>()?;
        let routing_parameter_set_id = cursor.read_fixed::<32>()?;
        let recent_position_floor = cursor.read_fixed::<32>()?;
        let action_balance_blinding = cursor.read_fr()?;
        let nk = cursor.read_fixed::<32>()?;
        let asset_path = cursor.read_merkle_path()?;
        let asset_position = cursor.read_u64()?;
        let asset_indexed_leaf = decode_indexed_leaf(&mut cursor)?;
        let asset_indexed_leaf_dk_pub_affine = cursor.read_point_affine()?;
        let asset_indexed_leaf_ring_pk_affine = cursor.read_point_affine()?;
        let is_regulated = cursor.read_bool()?;
        let regulated_precision = cursor.read_u8()?;
        let unregulated_precision = cursor.read_u8()?;
        let routing_as_of_height = cursor.read_u64()?;
        let routing_nonce = cursor.read_fixed::<32>()?;
        let sender_compliance_path = cursor.read_merkle_path()?;
        let sender_compliance_position = cursor.read_u64()?;
        let sender_d = cursor.read_fixed::<32>()?;
        let sender_status = cursor.read_fixed::<32>()?;
        let shared = NoteReshapeSharedNoteContextWitnessV6 {
            asset_id: cursor.read_fixed::<32>()?,
            diversified_generator_affine: cursor.read_point_affine()?,
        };
        if n_in as usize != family_id.input_count() || n_out as usize != family_id.output_count() {
            bail!(
                "{} witness shape mismatch: got {}x{}, expected {}x{}",
                family_id.label(),
                n_in,
                n_out,
                family_id.input_count(),
                family_id.output_count()
            )
        }
        let spends = (0..n_in)
            .map(|_| {
                decode_spend(
                    &mut cursor,
                    family_id.spec().input_padding
                        == crate::note_reshape::InputPaddingPolicy::SyntheticPrivate,
                )
            })
            .collect::<Result<Vec<_>>>()?;
        let outputs = (0..n_out)
            .map(|_| decode_output(&mut cursor))
            .collect::<Result<Vec<_>>>()?;
        let balance_commitment_affine = cursor.read_point_affine()?;
        let ak_affine = cursor.read_point_affine()?;
        cursor.finish("note reshape witness")?;

        Ok(Self {
            family_id,
            total_length,
            n_in,
            n_out,
            anchor,
            claimed_statement_hash,
            asset_anchor,
            compliance_anchor,
            routing_tag,
            routing_parameter_set_id,
            recent_position_floor,
            action_balance_blinding,
            nk,
            asset_path,
            asset_position,
            asset_indexed_leaf,
            asset_indexed_leaf_dk_pub_affine,
            asset_indexed_leaf_ring_pk_affine,
            is_regulated,
            regulated_precision,
            unregulated_precision,
            routing_as_of_height,
            routing_nonce,
            sender_compliance_path,
            sender_compliance_position,
            sender_d,
            sender_status,
            shared,
            spends,
            outputs,
            balance_commitment_affine,
            ak_affine,
        })
    }
}

fn encode_spend(
    buf: &mut Vec<u8>,
    spend: &NoteReshapeSpendWitnessV6,
    synthetic_private_padding: bool,
) -> Result<()> {
    if synthetic_private_padding {
        put_u32(buf, if spend.is_dummy { 1 } else { 0 });
        put_bytes(buf, &spend.dummy_nullifier_seed);
    }
    put_bytes(buf, &spend.nullifier);
    put_bytes(buf, &spend.spent_note_blinding);
    put_bytes(buf, &spend.spent_note_amount);
    put_bytes(buf, &spend.state_commitment_commitment);
    put_u64(buf, spend.state_commitment_position);
    encode_triple_path_32(buf, &spend.state_commitment_auth_path)?;
    put_bytes(buf, &spend.spend_auth_randomizer);
    encode_point_affine(buf, &spend.rk_affine);
    put_u8(buf, u8::from(spend.history_required));
    Ok(())
}

fn decode_spend(
    cursor: &mut BinaryCursor<'_>,
    synthetic_private_padding: bool,
) -> Result<NoteReshapeSpendWitnessV6> {
    let (is_dummy, dummy_nullifier_seed) = if synthetic_private_padding {
        let is_dummy = match cursor.read_u32()? {
            0 => false,
            1 => true,
            value => bail!("invalid note reshape input dummy flag {value}"),
        };
        (is_dummy, cursor.read_fixed::<32>()?)
    } else {
        (false, [0u8; 32])
    };
    Ok(NoteReshapeSpendWitnessV6 {
        is_dummy,
        nullifier: cursor.read_fixed::<32>()?,
        dummy_nullifier_seed,
        spent_note_blinding: cursor.read_fixed::<32>()?,
        spent_note_amount: cursor.read_fixed::<32>()?,
        state_commitment_commitment: cursor.read_fixed::<32>()?,
        state_commitment_position: cursor.read_u64()?,
        state_commitment_auth_path: cursor.read_triple_path_32()?,
        spend_auth_randomizer: cursor.read_fr()?,
        rk_affine: cursor.read_point_affine()?,
        history_required: cursor.read_bool()?,
    })
}

fn encode_output(buf: &mut Vec<u8>, output: &NoteReshapeOutputWitnessV6) {
    put_bytes(buf, &output.note_commitment);
    put_bytes(buf, &output.created_note_blinding);
    put_bytes(buf, &output.created_note_amount);
}

fn decode_output(cursor: &mut BinaryCursor<'_>) -> Result<NoteReshapeOutputWitnessV6> {
    Ok(NoteReshapeOutputWitnessV6 {
        note_commitment: cursor.read_fixed::<32>()?,
        created_note_blinding: cursor.read_fixed::<32>()?,
        created_note_amount: cursor.read_fixed::<32>()?,
    })
}
